plugins/cilium-cni/cmd/cmd.go

```diff
@@ -665,6 +681,13 @@ func (cmd *Cmd) Add(args *skel.CmdArgs) (err error) {
 			if err != nil {
 				return fmt.Errorf("unable to setup interface datapath: %w", err)
 			}
+		case ipamOption.IPAMDelegatedPlugin:
+			if ipam.IPV4.MasterMac != "" || ipam.IPV6.MasterMac != "" {
+				err = interfaceAdd(ipConfig, ipam.IPV4, conf)
```

Comment:
Why isn't the delegated plugin doing this? I guess we've tested this that this doesn't break the delegated Azure plugin, but I find it very weird that we are installing a rule on behalf of a delegated plugin, if the delegated plugin does have all the information needed to do it itself.

Or am I missing something and there is information here that the delegated plugin does not have? The only thing that comes to mind is the masquerading exception, but that code here does not populate the necessary exclusion CIDR, so that doesn't seem to be it either.

Reply:
@gandro , more contexts could be foun in https://github.com/cilium/cilium/issues/34604 

Routes and masq exception are controlled and should only be controlled by Cilium(main CNI) as following reasons:

- Cilium knowns all routes and masq info. It's better to controlled them in one component. 
- Pod virtual interfaces are yet not created when calling delegated plugin, so ingress routes(control traffic to Pod) has no chance to be installed.

For masquerading cidr, it's still controlled through Cilium "ipv4NativeRoutingCIDR' option.

Reply:
> Pod virtual interfaces are yet not created when calling delegated plugin, so ingress routes(control traffic to Pod) has no chance to be installed.

I'm not sure I follow. The routes installed by the call to `interfaceAdd` do not contain any interface name. Why do the interfaces need to exist?

Reply:
Ah. I guess you are talking about the per-endpoint routes (ingress). But those are not installed here. Those are not configured by the CNI, so I don't think this is relevant here.

`interfaceAdd` creates egress rules, steering traffic from a certain source IP over a certain host interface. I strongly believe this is something that the delegated plugin should configure itself, rather than relying on our EC2 logic (which we might also want to change if EC2 changes, which then would break all delegated plugins. I guess that is my main worry)

Reply:
@gandro you are right, interfaceAdd() is called to install egress routes, veth/netkit is not needed for installting egress routes.

The main idea is:
1. Control all routes only in Cilium is better than spliting part of them to another plugin(Delegated IPAM).
2. IPAM plugin is designed for IP allocation purpose. Not suitable for modifying system network configuration.
3. Cilium is actually already support multiple interfaces native routing. Which is mainly used in ENI mode. So we could leverage the existing capability, and just leave the IP allocation part to delegated IPAM plugin.

Reply:
I understand the desire to offload functionality to Cilium, but this adds some tight coupling between Cilium and the delegated IPAM plugin on a very very leaky abstraction boundary:

This here assumes that every delegated plugin that returns a host-level "Interfaces" object will want an Amazon EC2-ENI-style egress rule set up by Cilium. This might be true for Volcegine, but it's not true for example by the delegated Azure IPAM plugin. The Azure IPAM plugin does not seem to return an "Interfaces" object right now, but what if it ever does? Cilium would start to install egress EC2-style rules that would break the Azure plugin, because it doesn't work like that.

> IPAM plugin is designed for IP allocation purpose. Not suitable for modifying system network configuration.

I kindly disagree with this. Any cloud-provider based IPAM implementation will always require system network configuration changes, to teach the underlying fabric on how pod IPs are attached. Some component has to implement it. You are arguing that the plugin is the wrong place. I'm arguing that doing this inside Cilium is the wrong place, because it requires Cilium to understand what a certain delegated IPAM plugin requires and hard-code that cloud-providers specific network configuration. 

The whole idea behind delegated IPAM plugins is that Cilium does no longer have to implement cloud-provider specific logic. Yes, there are the existing ENI IPAM modes that do implement cloud-provider specific logic, but this comes at a high maintenance cost that we want to reduce. This is why we are pushing to move as much logic as we can out of Cilium and into the delegated plugin.

---

Having said that, I strongly disagree with this approach here where the Cilium CNI plugin interprets the delegated IPAM plugin result as "this plugin wants EC2-style egress routes") - if there is Cilium-specific logic that you need to build upon to implement the delegated plugin (e.g. obtaining the masquerading exceptions for the egress rules), then I think we are open to add endponts to the [Cilium API](https://docs.cilium.io/en/stable/api/#api-ref) to support such use-cases. That makes it explicit what the delegated plugin wants, rather than assuming something.

In general, I think it might be worthwhile to take a step back and have a real CFP document that presents the whole architecture you have in mind for the Volcengine support. For example: 
- What will the delegated IPAM plugin do?
- What Cilium-adjacent CRDs do you expect to implement?
- What functionality (e.g. ingress/egress routing) do you require of Cilium? How do you plan to invoke this functionality?

That way we can have a better shared understanding of what you expect of Cilium and we can see what kind of functionality we think Cilium can do, and what functionality we think should _not_ be handled by Cilium.

Reply:
> If the 3rd delegation IPAM plugin whish to customize the routes by themself, they cloud not return the uplink hints in the result.

As mentioned above, I'd prefer it the other way around: Let's have the delegated IPAM plugin call Cilium API if it wants things to be done by Cilium. That way, we can also provide as many Cilium-specific fields as we want, rather than relying information via the CNI result which was not designed for this purpose

Reply:
> Let's have the delegated IPAM plugin call Cilium API if it wants things to be done by Cilium

I don't thinks it is a good way to do such things, it would only bring even more coupling between Cilium and the delegated IPAM plugin. 

As @ruicao93 mentioned above, the delegated IPAM plugin should only do IP allocation and than tell main plugin what is the IP address of Pod and to which interface the pod supposed to bind as uplink, and don't care about in which way will main plugin realize traffic steering(using policy routing or bpf routing or event creating an ipvlan slave for pod).

Reply:
> I don't think it is a good way to do such things, it would only bring even more coupling between Cilium and the delegated IPAM plugin.

I'd argue the coupling is exactly the same. In both cases Cilium reacts to something that the delegated IPAM plug requested. The only difference is that the coupling is made explicit via an API designed for that purpose, rather than assuming that when a random delegated plugin returns host interfaces that wants a particular set of egress routing rules to be installed.

Reply:
I agree with @gandro's stance here. As the codeowners for the IPAM/cloud code, we are interested in reducing the amount of cloud-specific knowledge required to maintain the code because core Cilium developers don't necessarily have cloud-specific knowledge. While there are cloud IPAM implementations in the code today like ENI, it does not mean we want to open up Cilium for having all cloud-specific logic. Instead, it makes more sense to shift that maintenance to those with the expertise of whichever cloud via a delegated plugin. That plugin would reach out to Cilium API in order to retrieve whatever information it needs to carry out its job. It's an IPAM plugin for Cilium. This keeps the Cilium codebase simpler and less error-prone, while the vast majority of the complexity of the cloud/platform is kept contained inside the plugin.

Reply:
> when a random delegated plugin returns host interfaces that wants a particular set of egress routing rules to be installed.

My point is that routing is something ralated to the specific datapath which IPAM plugin is not supposed to be involved into. By passing uplink's information in CNI results, IPAM plugin can give cilium full control of datapath.

Reply:
i agree on @dakehero point but at same time don't want cilium cni to do custom stuff that fall outside of cni scope. Not sure if CNI/delegated ipam spec supports this @squeed ? if not, we have to standardize. 

@gandro I guess calling cilium api from delegated ipam would increase latency in pod startup time and bring in more complications.

Reply:
@tamilmani1989 it's allowed by CNI spec as squeed said in: https://github.com/cilium/cilium/pull/34779#issuecomment-2414088797

I'm requesting a change to append more description about the uplink in CNI spec:
https://github.com/containernetworking/cni/pull/1121